File: common/src/main/java/dev/mayaqq/estrogen/networking/messages/c2s/DashPacket.java
package dev.mayaqq.estrogen.networking.messages.c2s;

import com.teamresourceful.bytecodecs.base.ByteCodec;
import com.teamresourceful.resourcefullib.common.network.Packet;
import com.teamresourceful.resourcefullib.common.network.base.PacketType;
import com.teamresourceful.resourcefullib.common.network.base.ServerboundPacketType;
import com.teamresourceful.resourcefullib.common.network.defaults.CodecPacketType;
import dev.mayaqq.estrogen.Estrogen;
import dev.mayaqq.estrogen.features.dash.CommonDash;
import dev.mayaqq.estrogen.registry.EstrogenEffects;
import dev.mayaqq.estrogen.registry.EstrogenSounds;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.entity.player.Player;

import java.util.function.Consumer;

public record DashPacket(boolean sound) implements Packet<DashPacket> {

    public static final ServerboundPacketType<DashPacket> TYPE = new Type();

    @Override
    public PacketType<DashPacket> type() {
        return TYPE;
    }

    private static class Type extends CodecPacketType.Server<DashPacket> {

        public Type() {
            super(
                    DashPacket.class,
                    Estrogen.id("dash"),
                    ByteCodec.BOOLEAN.map(DashPacket::new, DashPacket::sound)
            );
        }

        @Override
        public Consumer<Player> handle(DashPacket message) {
            return (player) -> {
                if (player.hasEffect(EstrogenEffects.ESTROGEN_EFFECT.get()) && player.level() instanceof ServerLevel serverLevel) {
                    if (message.sound) serverLevel.playSound(null, player.blockPosition(), EstrogenSounds.DASH.get(), SoundSource.PLAYERS, 1.0F, 1.0F);
                    // dash cooldown
                    if (message.sound) CommonDash.setDashing(player.getUUID());
                    // summon particles around player
                    serverLevel.sendParticles(ParticleTypes.CLOUD, player.getX(), player.getY(), player.getZ(), 10, 0.5, 0.5, 0.5, 0.5);
                }
            };
        }
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/network/messages/c2s/DashPacket.kt
package dev.mayaqq.estrogen.network.messages.c2s

import dev.mayaqq.estrogen.content.EstrogenEffects
import dev.mayaqq.estrogen.content.EstrogenSounds
import dev.mayaqq.estrogen.content.particles.ColoredCloudParticleOptions
import dev.mayaqq.estrogen.content.particles.DashTrailParticleOptions
import dev.mayaqq.estrogen.features.dash.CommonDash
import dev.mayaqq.estrogen.utils.EstrogenColors
import invoke.kitty.kritter.registry.api.entry.holder
import kotlinx.serialization.Serializable
import net.minecraft.server.MinecraftServer
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.sounds.SoundSource
import kotlin.uuid.toKotlinUuid

@Serializable @JvmRecord
data class DashPacket(val isInitial: Boolean, val dashLevel: Int) {
    fun handle(server: MinecraftServer, sender: ServerPlayer) {
        if (sender.hasEffect(EstrogenEffects.Estrogen.holder)) {
            val level = sender.level() as ServerLevel

            val dashColor = EstrogenColors.getDashColor(dashLevel, true)

            // made them nicer hope u like maya c:
            level.sendParticles(
                ColoredCloudParticleOptions(dashColor, true),
                sender.x, sender.y + 0.5, sender.z, 3, 0.2, 0.7, 0.2, 0.18
            )

            if (isInitial) {
                // Set dashing and play dash sound if this is the initial packet
                CommonDash.setDashing(sender.uuid)
                level.playSound(null, sender, EstrogenSounds.DASH.get(), SoundSource.PLAYERS, 1.0f, 1.0f)
            } else {
                // Otherwise spawn trail
                level.sendParticles(
                    DashTrailParticleOptions(sender.uuid.toKotlinUuid(), dashColor),
                    sender.xOld, sender.yOld, sender.zOld,
                    0, 0.0, 0.0, 0.0, 0.0
                )
            }
        }
    }
}